0015-authority.md

```diff
@@ -0,0 +1,478 @@
+```csharp
+public override void OnGainedAuthority()
+{
+    Debug.Log("Authority Gained :)");
+}
+
+public override void OnLostAuthority()
+{
+    Debug.Log("Authority Lost :(");
+}
+```
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Why is Authority a mode?
+Why is authority not something like ownership which you give to a client. The reason for that is because authority is a relation between multiple peers. In an authoritative model a set of peers pass inputs to the authoritative peer for processing.
+
+Because we are a star topology we have some technical limitations. For instance we could not have one client generate the inputs but pass it to another client which holds authority (because this would break the encapsulation principle of time explained in the NetworkTime RFC).
+
+Because of the above we will only support authority models where the owner of an object collects inputs and sends them to the authoritative peer which has to be either the owner or the server.
+
+Accepting inputs from multiple peers is something we will not support with the current proposed design. The authority mode system allows for further extension to support those modes.
+
+## Ownership vs Authority
+This RFC will introduce some rules and responsibilities for each of the two systems.
+
+IsOwner:
+- Only one peer can own a NetworkObject.
+- The `owner` of a NetworkObject can use `ServerRpcs` on its `NetworkBehaviour` by default.
+- The `owner` is responsible for collecting inputs and redirecting them to the authority.
+- The `owner` is encouraged write to `NetworkVariables` with readpermission = authority but not to other NetworkVariables.
+- When the owner peer disconnects, the object it owns might get destroyed when `DontDestroyWithOwner` is set to false.
+
+HasAuthority:
+- The `authority` should run gameplay code such as physics, movement, path finding, animation etc.
+- The `authority` should update NetworkVariables which get used by other clients to visualize the game state.
+
+IsServer:
+- The server has permission to change ownership and authority of NetworkObjects.
+- `IsServer` should be avoided for gameplay related code.
+
+## NetworkObject Authority
+
+NetworkObject will get an Authority Property and a way to query for authority with `HasAuthority`:
+```csharp
+public Authority Authority => m_Authority;
+
+public bool HasAuthority
+{
+    get
+    {
+        if (IsSpawned) // This is needed because else we don't know whether we are a server.
+        {
+            return IsOwner ? Authority == Authority.Owner : NetworkManager.IsServer && Authority == Authority.Server;
+        }
+        throw new SpawnStateException($"Cannot get authority for not spawned {nameof(NetworkObject)}");
+    }
+}
+```
+
+There will also be a function to change authority:
+```csharp
+public void ChangeAuthority(Authority authority)
+{
+    Assert.IsFalse(authority == Authority.FromNetworkObject)
```

Comment:
If I want to put it back to FromNetworkObject after changing it to something else, why not allow it?

Reply:
Oh you are right, this is just a stupid pre-condition, ignore it 😄